common/trace.c

```diff
@@ -76,6 +76,39 @@ struct span {
 static struct span *active_spans = NULL;
 static struct span *current;
 
+/* Whether or not we're tracing things. We start in tracing mode, but
+ * may deactivate ourselves, most likely due to an API usage
+ * error. This is a sort of lame-duck mode, in which traces are
+ * no-ops. It is safer than crashing the entire daemon.
+ */
+static bool active = true;
+
+#define TRACE_ASSERT_RET(cond, msg, ret)                                       \
```

Comment:
Macros that return are an anti pattern: just have a macro for trace_assert_fail which hands #cond, file and line to bool trace_assert_fail_ and open code it. That keeps it readable.